jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,617 @@
+      , px_end_top_(empty_quote().first.as_integer()) {
+  }
+
+  /// The value used to represent an empty bid
+  static half_quote empty_bid() {
+    return half_quote(LOWEST_PRICE, 0);
+  }
+
+  /// The value used to represent an empty offer
+  static half_quote empty_offer() {
+    return half_quote(HIGHEST_PRICE, 0);
+  }
+
+  /// @returns an empty bid or offer based on compare function
+  /// empty bid for less, empty offer for greater.
+  half_quote empty_quote() const {
+    if (better_(price4_t(1), price4_t(0))) {
+      // buy side
+      return empty_bid();
+    }
+    // sell side
+    return empty_offer();
+  }
+
+  /// @returns the best bid price and quantity.
+  /// The inside is always at the top_levels_ (empty_quote.price when empty)
+  half_quote best_quote() const {
+    if (px_inside_ == empty_quote().first) {
+      return empty_quote();
+    }
+    int rel_px = price_to_relative(px_inside_);
+    return half_quote(px_inside_, top_levels_.at(rel_px));
+  }
+
+  /// @returns the worst bid price and quantity.
+  /// Check if the side is empty first (return empty_quote in this case)
+  /// The worst price is at the bottom_levels (if not empty),
+  /// at the top_levels otherwise.
+  half_quote worst_quote() const {
+    if (px_inside_ == empty_quote().first) {
+      return empty_quote(); // empty side
+    }
+    if (!bottom_levels_.empty()) {
+      // ... worst price is at the bottom_levels_
+      auto i = bottom_levels_.rbegin(); // get the worst bottom_levels price
+      return half_quote(i->first, i->second);
+    }
+    // ... worst price at the top_levels_
+    int rel_worst = relative_worst_top_level();
+    price4_t px_worst = relative_to_price(rel_worst);
+    return half_quote(px_worst, top_levels_.at(rel_worst));
+  }
+
+  /// @returns the number of levels with non-zero quantity for the order side.
+  /// This is the size of bottom_levels_ plus all non zero top_levels values
+  std::size_t count() const {
+    if (!bottom_levels_.empty()) {
+      return bottom_levels_.size() + top_levels_count();
+    }
+    return top_levels_count();
+  }
+
+  /**
+   * Add a price and quantity to the side order book.
+   *
+   * @param px the price of the new order
+   * @param qty the quantity of the new order
+   * @returns true if the inside changed
+   *
+   * Validates:
+   * - if px is off limits (<= empty_bid price or >= empty_offer price)
+   * - if qty <= 0
+   * - throws an exception if this is the case
+   * .
+   * Then:
+   * - if px is worse than the px_begin_top_ then px goes to bottom_levels
+   * - if px is better than the inside then changes inside, redefines limits
+   * if needed (moving the tail to bottom_levels if any)
+   * - finally, updates qty at the relative(px)
+   */
+  bool add_order(price4_t px, int qty) {
+    // verify that px and qty are valid values. Throws feed_error otherwise
+    if ((px <= LOWEST_PRICE) or (px >= HIGHEST_PRICE) or (qty <= 0)) {
```

Comment:
For what is worth, the priority of the comparison operators vs. the boolean operators is such that you do not need the parenthesis.  Leave them if you want to though.